mkosi/__init__.py

```diff
@@ -3717,13 +3722,19 @@ def calculate_bmap(args: CommandLineArguments, raw: Optional[BinaryIO]) -> Optio
 
 
 def save_cache(args: CommandLineArguments, root: str, raw: Optional[str], cache_path: Optional[str]) -> None:
-    if cache_path is None or raw is None:
+    if cache_path is None:
+        return
+
+    # bail out if cache already exists
+    if os.path.exists(cache_path):
         return
 
     with complete_step('Installing cache copy ',
                        'Successfully installed cache copy ' + cache_path):
 
         if args.output_format.is_disk_rw():
+            if raw is None:
+                return
```

Comment:
I'm not quite convinced that having the `if raw is None` check here communicates the intention very will, because if `raw` is `None` nothing will be done, but the messages will be printed. Maybe you could move this before the `with` statement as `if args.output_format.is_disk_rw() and raw is None`